### DISCUSSION TITLE
How does Data provider access CLDR data?

### DISCUSSION THREAD

**Discussion by riajain0412** [unicode-org/icu4x#2876] [Q&A] [answered]
I was exploring about data providers in ICU4X but I'm little bit confused regarding the process of Data Providers. According to my understanding, Data providers can read JSON file and then provide responses accordingly but where is that JSON file present? Do we have to create blob or what?

**Comment by sffc** [comment, marked as answer]
You need to convert CLDR data into a format ICU4X accepts at runtime. You have several choices: JSON, Postcard, and Databake. You might choose a different format depending on your needs.

`icu4x-datagen` is the tool that does this conversion.

You can read more in the data management tutorial here:

https://github.com/unicode-org/icu4x/blob/main/docs/tutorials/data_management.md

**Reply by riajain0412** [reply]
Okay, Got it. And also , I've one more doubt that can I use blob format data with icu_segmenter? When I try passing blob format data then it throws error so just wanted to check that other then icu_testdata how can I use generated ICU data with segmenter?

**Reply by Manishearth** [reply]
Yes, you can, you need to make sure your blob data contains all of the keys you need for segmenter.

**Reply by riajain0412** [reply]
Do you have any sample code for this?

**Reply by riajain0412** [reply]
I'm running the below code but this is throwing this error -> the trait `for<'de> Deserialize<'_>` is not implemented for `RuleBreakDataV1<'de>

`use icu_segmenter::LineBreakSegmenter;

use icu::segmenter::LineSegmenter;

use icu_testdata;

use icu_provider::prelude::*;

use icu_provider_blob::BlobDataProvider;

use icu_segmenter::provider::LineBreakDataV1Marker;

use icu_segmenter::provider::RuleBreakDataV1;

use ::serde::{de, Deserialize};

fn main() {
let blob = std::fs::read("my-data-blob").expect("Failed to read file");
    let buffer_provider = BlobDataProvider::try_new_from_blob(blob.into_boxed_slice()).expect("Failed to initialize Data Provider.");
    let provider = LineBreakDataV1Marker();
    let payload = DataProvider::<LineBreakDataV1Marker>::load(
            &icu_testdata::buffer().as_deserializing(),
            Default::default(),
        )
        .expect("Loading should succeed!")
        .take_payload()
        .expect("Data should be present!");
}`

**Reply by Manishearth** [reply]
You have to build with the serde feature enabled

**Reply by riajain0412** [reply]
Below are the dependencies which I have added in my Cargo.toml file :
[dependencies]
icu = {version= "1.0.0", features = ["serde", "experimental"]}
icu_provider_blob = "1.0.0"
icu_provider = "1.0.0"
icu_segmenter = "0.7.0"
serde = { version = "1.0.101", optional = true, features = ["derive"] }
serde_json = "*"
serde_macros = { version = "*", optional = true }
icu_testdata = {version= "1.0.0", features = ["icu_segmenter", "buffer"]}
unicode-segmentation = "1.0.0"

**Reply by Manishearth** [reply]
Ah you also need to enable serde on icu_segmenter

In future versions this will be automatic

**Reply by riajain0412** [reply]
Okay, Adding serde in icu_segmenter resolved the above error. Now my code looks something like this:
But I'm facing some new error due to last line. Is the code what I've written is correct? 2 errors are coming:
1. the trait `icu_provider::DataProvider<LineBreakDataV1Marker>` is not implemented for `&RuleBreakDataV1<'_>`the trait `icu_provider::DataProvider<LineBreakDataV1Marker>` is not implemented for `&RuleBreakDataV1<'_>`
2. the trait `icu_provider::DataProvider<UCharDictionaryBreakDataV1Marker>` is not implemented for `&RuleBreakDataV1<'_>`

let blob = std::fs::read("my-data-blob").expect("Failed to read file");
    let buffer_provider = BlobDataProvider::try_new_from_blob(blob.into_boxed_slice()).expect("Failed to initialize Data Provider.");
    // let provider = LineBreakDataV1Marker();
    let payload = DataProvider::<LineBreakDataV1Marker>::load(
            &buffer_provider.as_deserializing(),
            Default::default(),
        )
        .expect("Loading should succeed!")
        .take_payload()
        .expect("Data should be present!");
    // let segmenter = LineBreakSegmenter::try_new_unstable(&buffer_provider).expect("Data exists");
    let lb_data: &RuleBreakDataV1 = payload.get();
    let segmenter = LineBreakSegmenter::try_new_unstable(&lb_data).expect("Data exists");

**Reply by Manishearth** [reply]
Use `try_new_with_buffer_provider`, or call `as_deserializing()` before passing to `try_new_unstable`

**Reply by riajain0412** [reply]
After changing try_new_unstable to try_new_with_buffer_provider, it is giving another error that:
The trait `BufferProvider` is not implemented for `&RuleBreakDataV1<`_>

**Reply by riajain0412** [reply]
I tried this with blob data as well. But when I do `cargo run` it says segmenter key is missing but while generating blob data I passed --all-keys argument

**Reply by riajain0412** [reply]
@Manishearth All I want is to implement LineBreakSegmenter without icu_testdata i.e. I want to build my own data and then use it for segmentation. But I'm confused how to achieve that. And also, as segmentation is experimental as of now so do you have segmenter key or not?

**Reply by sffc** [reply]
You need to enable all the serde features you need. On the docs page for `LineBreakSegmenter::try_new_with_buffer_provider`, you can see the note: **✨ Enabled with the "serde" feature.**

https://icu4x.unicode.org/doc/icu_segmenter/struct.LineSegmenter.html#method.try_new_with_buffer_provider

The "serde" feature can be enabled on icu_segmenter or icu. For example:

```
icu_segmenter = { version = "0.7.0", features = ["serde"] }
```

**Reply by sffc** [reply]
In general, whenever there is an error along the lines of "trait X is not implemented for Y", much of the time it means that you haven't enabled the right features. Here's an issue to follow up with improving the developer experience around this: https://github.com/unicode-org/icu4x/issues/2853

**Reply by riajain0412** [reply]
Hi @sffc, I had already added serde feature in icu_segmenter. Error which I'm getting now is `{kind: MissingDataKey, key: Some(DataKey{segmenter/line@1}), str_context: None }`. So I was wondering that why is this key missing? I ran icu4x-datagen for all-keys. Is this because segmenter is a experimental feature?

**Reply by Manishearth** [reply]
You have to build icu4x-datagen in experimental mode as well.

**Reply by sffc** [reply]
To generate your postcard file, please run: `icu4x-datagen --icuexport-tag latest --format blob --keys "segmenter/line@1" --all-locales -o segmenterline.postcard`

You need to run it with the "experimental" and "bin" features enabled on icu_datagen

**Reply by sffc** [reply]
To include LSTM and Dictionary data in your line break segmenter: `icu4x-datagen --icuexport-tag latest --format blob --keys "segmenter/line@1" "segmenter/lstm@1" "segmenter/dictionary@1" --all-locales -o segmenterline.postcard`

**Reply by riajain0412** [reply]
> You have to build icu4x-datagen in experimental mode as well.

How to do that?

**Reply by riajain0412** [reply]
> To generate your postcard file, please run: `icu4x-datagen --icuexport-tag latest --format blob --keys "segmenter/line@1" --all-locales -o segmenterline.postcard`
> 
> You need to run it with the "experimental" and "bin" features enabled on icu_datagen

What is the use of postcard file? Is this required even if I'm creating blob?

**Reply by Manishearth** [reply]
That _is_ the blob

**Reply by sffc** [reply]
`cargo install --features experimental,bin icu_datagen`

**Reply by riajain0412** [reply]
Okay, This finally worked. Thank you so much @sffc @Manishearth . 
I've one last query that I was now trying WordBreakSegmenter but this time I only included "segmenter/word@1" and "segmenter/dictionary@1" but this gave incorrect results. I have not mentioned feature lstm anywhere. I want to use dictionary based segmentation.

**Reply by sffc** [reply]
With icu_segmenter version 0.7, if the lstm feature is off and your data file contains "segmenter/dictionary@1" data, yhou should get dictionary segmentation.

Note that the lstm feature could get implicitly enabled if you have a dependency on something that enables it, such as icu_datagen with the experimental feature. You don't need icu_datagen in your Cargo.toml file unless you run it in your build.rs. This is one of the issues we want to clean up before stabilizing icu_segmenter; see https://github.com/unicode-org/icu4x/issues/2829

**Reply by riajain0412** [reply]
Okay, yes, removing icu_datagen from Cargo.toml worked. Thank you so much for your help.

**Reply by riajain0412** [reply]
Have you guys done any benchmark testing for segmentation wrt ICU4C? I want to know that how ICU4X is better than ICU4C in terms of segmentation.
